tests/ga/test_extension.py

```diff
@@ -2322,324 +2324,320 @@ def test_it_should_parse_boolean_values_encoded_as_strings(self):
 
-            self.assertTrue('-update' in kwargs['command'] and ExtCommandEnvVariable.DisableReturnCode in kwargs['env'],
-                            "The update command should have Disable Failed in env variable")
+                    # Note that if installCommand is supposed to fail, this will erroneously raise.
+                    (first_ext, ExtensionCommandNames.ENABLE)
+                )
 
-    def test_uninstall_failed_env_variable_should_set_for_install_when_continue_on_update_failure_is_true(
-            self, *args):
-        old_handler_i = self._get_ext_handler_instance('foo', '1.0.0')
-        old_handler_i.set_handler_state(ExtHandlerState.Enabled)
-        new_handler_i = self._get_ext_handler_instance('foo', '1.0.1', continue_on_update_failure=True)
+            protocol.mock_wire_data.set_extensions_config_version("1.1.0")
+            protocol.mock_wire_data.set_incarnation(2)
+            protocol.client.update_goal_state()
+
+            with enable_invocations(first_ext, upgraded_ext) as invocation_record:
+                exthandlers_handler.run()
+            
+                return invocation_record
 
-        with patch.object(CGroupConfigurator.get_instance(), "start_extension_command",
-                          side_effect=['ok', 'ok', ExtensionError('uninstall Failed'), 'ok']) as patch_start_cmd:
 
-            ExtHandlersHandler._update_extension_handler_and_return_if_failed(old_handler_i, new_handler_i)
+    def test_non_enabled_ext_should_not_be_disabled_at_ver_update(self):
 
-            args, kwargs = patch_start_cmd.call_args
+        first_ext = extension_emulator(enable_action=Actions.fail_action)
+        second_ext = extension_emulator(version="1.1.0")
+        
+        invocation_record = TestExtensionUpdateOnFailure._do_upgrade_scenario(first_ext, second_ext)
+
+        invocation_record.compare(
+            (second_ext, ExtensionCommandNames.UPDATE),
+            (first_ext, ExtensionCommandNames.UNINSTALL),
+            (second_ext, ExtensionCommandNames.INSTALL),
+            (second_ext, ExtensionCommandNames.ENABLE)
+        )
+
+    def test_disable_failed_env_variable_should_be_set_for_update_cmd_when_continue_on_update_failure_is_true(self):
+        exit_code, disable_action = Actions.generate_unique_fail()
+
+        first_ext = extension_emulator(disable_action=disable_action)
+        second_ext = extension_emulator(version="1.1.0", continue_on_update_failure=True)
+
+        invocation_record = TestExtensionUpdateOnFailure._do_upgrade_scenario(first_ext, second_ext)
+
+        invocation_record.compare(
+            (first_ext, ExtensionCommandNames.DISABLE),
+            (second_ext, ExtensionCommandNames.UPDATE),
+            (first_ext, ExtensionCommandNames.UNINSTALL),
+            (second_ext, ExtensionCommandNames.INSTALL),
+            (second_ext, ExtensionCommandNames.ENABLE)
+        )
+
+        _, kwargs = second_ext.actions[ExtensionCommandNames.UPDATE].call_args
```

Comment:
From the ExtensionEmulator class, `self.actions` is a dict of actions. And according to `Actions` class, the functions are just plain old objects, not Mocks. 

```python
self.actions = {
            ExtensionCommandNames.INSTALL: install_action,
            ExtensionCommandNames.UNINSTALL: uninstall_action,
            ExtensionCommandNames.UPDATE: update_action,
            ExtensionCommandNames.ENABLE: enable_action,
            ExtensionCommandNames.DISABLE: disable_action
        }

class Actions(object):
    """
    A collection of static methods providing some basic functionality for the ExtensionEmulator
    class' actions.
    """

    @staticmethod
    def succeed_action(*args, **kwargs):
        """
        A nop action with the correct function signature for ExtensionEmulator actions.
        """
        return 0
```

In the patched_popen, you're just calling the `Action` of the emulator directly with whatever arguments are being passed to it - 
```python
return next(
                emulator.actions[ExtensionCommandNames(command_name)]
                for emulator in emulators
                if emulator.matches(ext_name, ext_version)
            )(cmd, *args, **kwargs)
```

My main question is how is this line working if this is not a mock? I feel like I'm missing something
```python
_, kwargs = second_ext.actions[ExtensionCommandNames.UPDATE].call_args
```

Reply:
Sorry, this is definitely something that could be more explicit in the code, looking back over it.

The thing you're missing is the private function `_extend_func` in `extension_emulator.py`. This is called within the factory method (`extension_emulator` in that same file), and wraps not only the result of the passed function in a mock popen object, but also the function itself in a mock. It's near the bottom of the page, if I remember correctly.

I think the lack of clarity is actually pointing to a issue with how the code is written, though; I don't think it's ideal to call `_extend_func` within only the factory method. Do you have any thoughts on alternatives?

Could it be as simple as moving the `_extend_func` call into the `ExtensionEmulator` class and better documenting it? I think it would be nice if it was relatively opaque to the user of the test (but clear enough if someone is to dive into the `ExtensionEmulator` class), because that way, a regular lambda can be used, and the test writer wouldn't have to care about the necessary massaging to get that lambda into the correct format.

Reply:
Ahh I see it now! I definitely missed that one lol. And yes I agree with your idea, it would be better to add it to the `ExtensionEmulator` class and make the `_extend_func` an internal lambda so that anyone reading the code can understand what's happening there. 
Also while you're at it, make sure to rename the function to something more specific like - `_extend_action_to_popen_mock`  or something.